### PR TITLE
Remove operation commands, replacing them by apply-operations

### LINKED ISSUE #5539: Remove backend commands exposed for each operation
So far, each operation that the user can run on a project comes with the following Java classes in the backend:
* an Operation class, which holds the metadata for the operation and is responsible for its JSON serialization (which is exposed in the history tab, among others)
* a Change class (often reused by different operations), which is responsible for actually applying the operation to the project (carrying out the corresponding transformation)
* a Command class, which exposes an HTTP API to initiate the operation on a project.

Therefore, each operation comes with its own HTTP endpoint to apply it, and the frontend can call that endpoint when the user clicks on some menu item or validates some dialog, for instance.

At the same time, we have a generic endpoint to apply a series of operations, represented by their JSON serialization, on a given project. This is used in the history tab, to let the user apply a workflow exported in JSON.

This endpoint (called `apply-operations`) makes all the other operation-specific commands redundant: we should be able to use it to apply any single operation, hence removing the need for a fairly large number of other commands.

This would have the following benefits:
* remove a lot of Java classes
* refactor the frontend so that making any change to the project goes through the same code, making it easier to implement logic that should apply to all those changes (such as warning the user that they are discarding a part of the history if there are "future" history entries: #3184)
* make it easier to implement new operations, by removing the need to implement a corresponding command (but the move can be done without any breaking change with respect to the extension interface - extensions are still free to define their own commands for each operation they define).

This will likely require improving the error handling in the `apply-operations` endpoint, since it is likely that operation-specific endpoints currently offer a better handling. When an operation's JSON representation is incomplete or invalid, we want to be able to return an appropriate response which should be replicated in the frontend accordingly. This will benefit the "Apply operations" UX as well.

### PR DESCRIPTION
**Pull request by wetneb**
Fixes #5539.

This builds on #5554, so the diff will become slightly cleaner once that one is merged.

### CODE REVIEW

**Comment by wetneb** [comment]
It's unclear to me why the CI does not run, probably because I opened the PR on master and not 4.0 in the first place. It should run again when #5554 is merged and I rebase this on top of 4.0.

File: main/webapp/modules/core/scripts/views/data-table/menu-reconcile.js [hunk 1]
```diff
@@ -132,17 +136,18 @@ DataTableColumnHeaderUI.extendMenu(function(column, columnHeaderUI, menu) {
...
-        Refine.postCoreProcess(
-        "recon-match-specific-topic-to-cells",
+        Refine.postOperation(
         {
+            op: "core/recon-match-specific-topic-to-cells",
             columnName: column.name,
-            topicID: data.id,
-            topicName: data.name,
-            types: types.join(","),
+            match: {
```
Comment (tfmorris) [inline main/webapp/modules/core/scripts/views/data-table/menu-reconcile.js:143]: Are the parameters changing here? It's hard to tell from the diff without digging into the old function.

**Review by tfmorris — COMMENTED** [review]
Between the size of these changes and my lack of familiarity with the 4.x proto, it's not really possible for me to provide effective feedback. I took a quick skim through and made comments on a few things that caught my eye.

For something this size, some reviewer notes would be useful, even if brief. The biggest questions in my mind concern compatibility. Does it break Javascript extensions? Is it an incompatible protocol/REST API change? If you're making an incompatible change to the REST API, is this the right design for the new API? 

One way to effectively summarize the changes would be to include migration instructions for extension developers, API client developers, etc.

Sorry I can't provide more effective feedback. The compatibility discussion is something that might be a useful topic for the developers mailing list.

File: main/src/main/java/org/openrefine/commands/history/ApplyOperationsCommand.java [hunk 2]
```diff
@@ -102,29 +102,32 @@ protected OperationApplicationResult applyOperation(Project project, JsonNode op
...
+                applicationResult = new OperationApplicationResult();
             }
-        } else {
-            return new OperationApplicationResult("Operation could not be parsed");
+            return applicationResult;
+        } catch (Exception e) {
+            // TODO make catch block narrower, only catching certain expected exceptions
+            // such as DoesNotApplyException. This requires tightening the exceptions in the Process
+            // interface too.
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (wetneb) [reply main/src/main/java/org/openrefine/commands/history/ApplyOperationsCommand.java:129]: Good catch!

**Comment by wetneb** [comment]
Thanks for having a look Tom!
Sorry for the lack of context, I should have pinged you on the issue instead: #5539.
Have you had a look at it already?

In terms of breaking changes:
* it removes a lot of internal API endpoints, but as [documented](https://openrefine.org/docs/technical-reference/openrefine-api) we do not offer stability guarantees for external uses;
* those commands are still something extensions could legitimately make use of, so you are right that this should be mentioned in a migration guide for extension developers. This is the main reason why I am proposing this change in 4.x and not in master (this change is actually independent from other architecture changes in 4.x), in the interest of keeping 3.x as breakage-free.

Overall there are quite a few other changes to mention in such a guide, and I am in the process of writing it. For a long time I have been wondering whether it is reasonable to write a migration guide at all given the scale of the changes between 3.x and 4.x, but I think this is actually manageable and will be useful.

Here is a draft of an entry about this change in such a guide.

## Removal of operation commands

So far, each Operation in OpenRefine had a Command counterpart. This Command exposed an HTTP endpoint to apply the Operation on a given project. We have removed such commands, as the same effect can be obtained using the `command/core/apply-operations` endpoint, which can be used to apply any operation defined by its JSON serialization.

For instance:

In 3.x, you could apply a "mass edit" operation in two ways:
1. using the `command/core/mass-edit` HTTP endpoint, by passing the parameters of the operation as GET and POST parameters:
```
curl 'http://127.0.0.1:3333/command/core/mass-edit?project=2139117096609' --data-raw 'columnName=column+A&expression=value&edits=%5B%7B%22from%22%3A%5B%22e%22%5D%2C%22to%22%3A%22new+value%22%2C%22type%22%3A%22text%22%7D%5D&engine=%7B%22facets%22%3A%5B%5D%2C%22mode%22%3A%22row-based%22%7D&csrf_token=1VUQgm2192SyszMp6bTJMAVabvfH9LRe'
```
2. using the `command/core/apply-operations` HTTP endpoint, by passing the JSON serialization of the operation, wrapped in an array, as a POST parameter
```
curl 'http://127.0.0.1:3333/command/core/apply-operations?project=2129502021603' --data-raw 'operations=%5B%7B%22op%22%3A%22core%2Fmass-edit%22%2C%22columnName%22%3A%22column%22%2C%22expression%22%3A%22value%22%2C%22edits%22%3A%22%5B%7B%5C%22from%5C%22%3A%5B%5C%22knot%5C%22%5D%2C%5C%22to%5C%22%3A%5C%22not%5C%22%2C%5C%22type%5C%22%3A%5C%22text%5C%22%7D%5D%22%2C%22engineConfig%22%3A%7B%22facets%22%3A%5B%5D%2C%22mode%22%3A%22row-based%22%2C%22aggregationLimit%22%3A1000%7D%7D%5D&engine=%7B%22facets%22%3A%5B%5D%2C%22mode%22%3A%22row-based%22%2C%22aggregationLimit%22%3A1000%7D&csrf_token=ZwLqOPVYbNPveWp15PGMArMUCJG4goy8'
```

In 4.x, we have removed the first option and changed the frontend so that it uses the second option instead.

In Javascript code, you are invited to migrate your calls to core operations away from the `Refine.postCoreProcess` method, and use `Refine.postOperation` instead. It takes the following arguments:
* the representation of the operation configuration as a Javascript object (which will be serialized in JSON)
* the update options for the grid, as in `Refine.postCoreProcess`
* the callbacks to call after the operation was applied, or in case of failure, as in `Refine.postCoreProcess`

You can also use this syntax with the operations defined in your own extension (if any).
You can continue to use any command defined by your extension with `Refine.postProcess` if you wish.

File: main/webapp/modules/core/scripts/views/data-table/menu-reconcile.js (hunk 1 above)
Comment (wetneb) [reply main/webapp/modules/core/scripts/views/data-table/menu-reconcile.js:143]: See the explanation below: this change comes from the difference between the encoding of the operation parameters as POST parameters of the operation command, and as a JSON object in the internal representation of the operation.

File: main/webapp/modules/core/scripts/views/data-table/menu-reconcile.js [hunk 3]
```diff
@@ -309,32 +313,31 @@ DataTableColumnHeaderUI.extendMenu(function(column, columnHeaderUI, menu) {
     elmts.cancelButton.on('click',dismiss);
     elmts.okButton.on('click',function() {
       var config = {
+        op: "core/recon-copy-across-columns", 
         fromColumnName: column.name,
-        toColumnName: [],
-        judgment: [],
+        toColumnNames: [],
```
[the remark this answers is not in the document]
Comment (wetneb) [reply main/webapp/modules/core/scripts/views/data-table/menu-reconcile.js:318]: See the explanation below: this change comes from the difference between the encoding of the operation parameters as POST parameters of the operation command, and as a JSON object in the internal representation of the operation (which remains unchanged).

**Comment by tfmorris** [comment]
Thanks for the background. I had gone back to look at the issue, but the additional clarifications help.

Some final comments:
* "Do the things described by this JSON blob" feels a little weird from a RESTful API design point of view. I guess how important that is depends on whether this is also intended to be the public API (or a step towards it) or a purely internal API. Sounds expensive to maintain two separate HTTP APIs. If I had the time, I'd probably look at what other APIs are doing in this space from a design style point of view.
* Similarly, if you're changing things in a breaking fashion, it'd be worth making the new naming schemes internally consistent (singular vs plural, verb vs noun, etc). It's difficult to judge as part of a diff because it's not easy to check anything which didn't change to make sure it's consistent.

While I understand the value in doing things incrementally, it has the inherent problem that it makes it hard to evaluate whether an increment is moving toward some undefined future end point. Things like APIs really need to be reviewed as a whole to see whether they're internally consistent and hang together holistically. The current API is very likely not internally consistent because it was developed incrementally over time. But perhaps that's ok too.

**Comment by wetneb** [comment]
Yes, for restfulness of this API, there is still a lot to do.

One first step, for the `apply-operations` endpoint, would be to POST the JSON payload directly (with header `content-type: application/json`), without embedding it URL-encoded in form parameters. We still need to pass the project ID and CSRF token though, so it's not that simple.

I have no plans to push for turning this API into a more public-facing interface, so from that perspective I do not see a big urge to make it perfectly RESTful. That being said, there are a couple of things going in that direction that I would like to do:
* making sure the content-type headers of all responses are correctly set (for a bunch of commands, we get errors in the browser console because the browser tries to parse JSON as XML)
* making sure we use HTTP codes appropriately (it is very common for a command to return an error with HTTP 200 code).

With this PR, the only naming schemes remaining are those of the operations, so I would rather leave those untouched for now. So far, JSON representations of sequences of operations generated by 3.x will be understood by 4.x, and although I do not thing this is a critical feature to maintain, I would not break it just for the sake of normalizing naming. We can reconsider that when we need to introduce changes to those representations.

**Pull request merged**